### PR TITLE
Feature/add ireq model

### PR DESCRIPTION
**Pull request by ChuanLi32781**
Add IREQ  model based on ISO 11079

## Summary by CodeRabbit

* **New Features**
  * Added ISO 11079 Required Clothing Insulation (IREQ) and Duration Limited Exposure (DLE) calculations and made them available via the public API.

* **Documentation**
  * Added a dedicated IREQ/DLE documentation section and corrected several punctuation/formatting and encoding issues across docs and changelog.

* **Examples**
  * Added an example script demonstrating scalar and array-based usage.

* **Tests**
  * Added tests covering scalar/array results, broadcasting, clamping, validation and public API exposure.

### CODE REVIEW

File: docs/documentation/models.rst [hunk 1]
```diff
@@ -20,7 +20,7 @@ Below are some key characteristics and criteria of the adaptive model, according
   *   There is no mechanical cooling or heating system in operation.
   *   Occupants have metabolic rates ranging from 1.0 to 1.5 met.
   *   Occupants are able to adjust their clothing to indoor or outdoor thermal conditions within a range of 0.5 to 1.0 clo.
-  *   The prevailing mean outdoor temperature is between 10°C (50°F) and 33.5°C (92.3°F).
+  *   The prevailing mean outdoor temperature is between 10��C (50��F) and 33.5��C (92.3��F).
```
Comment (FedericoTartarini) [inline docs/documentation/models.rst:23]: Could you please explain this change to me?

File: pythermalcomfort/models/__init__.py [hunk 2]
```diff
@@ -36,7 +36,7 @@
 from .work_capacity_hothaps import work_capacity_hothaps
 from .work_capacity_iso import work_capacity_iso
 from .work_capacity_niosh import work_capacity_niosh
-
+from .ireq import calc_ireq
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/__init__.py:39]: None of the model is called calc_XXX We need to change the name and call it ireq

File: pythermalcomfort/models/ireq.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import numpy as np
+from typing import Union, Dict, Any, List
+
+from pythermalcomfort.classes_input import IREQInputs
+from pythermalcomfort.classes_return import IREQResult
+
+def calc_ireq(M: Union[float, list, np.ndarray],
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py:7]: See my previous comment about the function name.

File: pythermalcomfort/models/ireq.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        
+        result = calc_ireq(M=175.0, W=0.0, ta=-15.0, tr=-15.0, p=50.0, w=1.1, v=2.0, rh=55.0, clo=2.8)
+        print(result.IREQminimal)
+
+    References
+    ----------
+    ISO 11079:2007 Standard
+    """
+    # Check if inputs are all scalar to correctly format the output later
+    is_scalar = np.isscalar(M) and np.isscalar(W) and np.isscalar(ta) and \
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py:73]: Please reformat the file using ruff

File: pythermalcomfort/models/ireq.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
+import numpy as np
+from typing import Union, Dict, Any, List
+
+from pythermalcomfort.classes_input import IREQInputs
+from pythermalcomfort.classes_return import IREQResult
+
+def calc_ireq(M: Union[float, list, np.ndarray],
+              W: Union[float, list, np.ndarray],
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py:8]: We never use variable names which are capitalized because they are not PyTonic. Can you please review how we name the other variables in the other function and use the same approach?

File: pythermalcomfort/models/ireq.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+              ta: Union[float, list, np.ndarray],
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py:9]: We never call this ta but we call it tdb. Please review the other functions and use the same naming convention.

File: CHANGELOG.rst [hunk 7]
```diff
@@ -33,9 +34,9 @@ Changelog
     - removed outputs: ``water_loss``, ``water_loss_watt``
     - added outputs (with units):
 
-        * ``sweat_loss_g`` [g] — cumulative sweat mass per person (not area‑normalised)
-        * ``sweat_rate_watt`` [W·m⁻²] — instantaneous evaporative heat flux at skin
-        * ``evap_load_wm2_min`` [W·min·m⁻²] — accumulated evaporative load for chaining
+        * ``sweat_loss_g`` [g] -cumulative sweat mass per person (not area-normalised)
+        * ``sweat_rate_watt`` [W/m²] -instantaneous evaporative heat flux at skin
```
Comment (FedericoTartarini) [inline CHANGELOG.rst:38]: Please only change the files that are related to the IREQ You can submit then another pull request in which you propose to fix this, if needed.

File: CHANGELOG.rst [hunk 8]
```diff
@@ -105,7 +106,7 @@ Changelog
 3.1.0 (2025-04-28)
 -------------------
 * Updated the PHS model in compliance with the ISO 7933:2023 standard
-    - Added default‑kwarg overrides for 2023 mode (f_r, t_re, t_cr_eq)
+    - Added default-慿warg overrides for 2023 mode (f_r, t_re, t_cr_eq)
```
Comment (FedericoTartarini) [inline CHANGELOG.rst:109]: Be very careful that you don't introduce wrong characters as here, that you have introduced a Chinese character. You should also not change the change log unless it is related to this PR If you spot errors you can submit a separate pull request.

**Review by FedericoTartarini — CHANGES_REQUESTED** [review]
@ChuanLi32781 Great work, thank you very much for working on this, I am very sorry that I didn't notice that you submitted this poor request, apologies for that. I've left a few comments which I would like you to carefully review. Also please note that CodeRabbit has left some comments. Once you have addressed them all, please send me a message so I can do a second review.

File: docs/documentation/models.rst [hunk 9]
```diff
@@ -131,14 +131,14 @@ Gagge-Ji two-node model
...
+For instance, at 80% relative humidity, the heat index is only valid within a temperature range of 288–304 K.
 To address this issue, Lu and Romps (2022) [lu]_ developed a new model that extends the range of validity of the heat index.
 
-pythermalcomfort therefore includes two equations to calculate the Heat Index. One in accordance with the new Lu and Romps (2022) model which is an extension of the first version of Steadman's (1979) apparent temperature :py:func:`~pythermalcomfort.models.heat_index_lu.heat_index_lu`.
-The other is developed by Rothfusz (1990) and it is a simplified model derived by multiple regression analysis in temperature and relative humidity from the first version of Steadman's (1979) apparent temperature (AT) [Rothfusz1990]_ :py:func:`~pythermalcomfort.models.heat_index_rothfusz.heat_index_rothfusz`.
+pythermalcomfort therefore includes two equations to calculate the Heat Index. One in accordance with the new Lu and Romps (2022) model which is an extension of the first version of Steadman’s (1979) apparent temperature :py:class:`~pythermalcomfort.models.heat_index_lu.heat_index_lu`.
```
Comment (FedericoTartarini) [inline docs/documentation/models.rst:140]: I'm not entirely sure why you changed these because they are not related to the IREQ model. Could you please explain?

File: examples/ireq.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
+from pythermalcomfort.models import ireq
```
Comment (FedericoTartarini) [inline examples/ireq.py]: No need to add this file, we can add the example in the doc string of the model. Please delete this file.

File: pythermalcomfort/models/ireq.py [hunk 11]
```diff
@@ -1,227 +1,388 @@
...
-        
-        result = calc_ireq(M=175.0, W=0.0, ta=-15.0, tr=-15.0, p=50.0, w=1.1, v=2.0, rh=55.0, clo=2.8)
-        print(result.IREQminimal)
+    IREQ
+        A dataclass containing ``ireq_min``, ``ireq_neutral``, ``icl_min``,
+        ``icl_neutral``, ``dle_min``, and ``dle_neutral``.
 
     References
     ----------
     ISO 11079:2007 Standard
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py]: Please add proper reference as we have added for the other model. See for example the PMV model as an example.

File: pythermalcomfort/models/ireq.py [hunk 12]
```diff
@@ -1,227 +1,388 @@
...
-    Pex = 0.1333 * np.exp(18.6686 - 4030.183 / (Tex + 235.0))
-    Pa = (rh_val / 100.0) * 0.1333 * np.exp(18.6686 - 4030.183 / (ta_val + 235.0))
-    
-    ArAdu = 0.77
-    results_dict = {}
+    ar_adu = 0.77
+    results = {}
 
-    for calculation in [1, 2]:
+    for calculation in (1, 2):
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py]: It is not very clear to me what you are doing in here, why you are looping through a tuple and why this tuple is made of one and two.

File: pythermalcomfort/models/ireq.py [hunk 13]
```diff
@@ -1,227 +1,388 @@
...
     )
+
+
+def _clothing_constant_part(p, v, walk_sp):
+    return (
+        0.54 * np.exp(-0.15 * v - 0.22 * walk_sp) * (p**0.075) - 0.06 * np.log(p) + 0.5
+    )
+
+
+def _valid_iso_11079_inputs(met, tdb, v, walk_sp):
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py]: We need to improve the doc string of the model and mention these inputs over there. Please have a look as reference to the UTCI or PMV function.

File: tests/test_ireq.py [hunk 14]
```diff
@@ -1,121 +1,182 @@
 import numpy as np
 import pytest
 
-from pythermalcomfort.models.ireq import calc_ireq
+from pythermalcomfort.classes_return import IREQ
```
Comment (FedericoTartarini) [inline tests/test_ireq.py:4]: Have you included in this test file all the tests from the ISO standard?

File: CHANGELOG.rst [hunk 15]
```diff
@@ -34,9 +38,9 @@
     - removed outputs: ``water_loss``, ``water_loss_watt``
     - added outputs (with units):
 
-        * ``sweat_loss_g`` [g] -cumulative sweat mass per person (not area-normalised)
-        * ``sweat_rate_watt`` [W/m²] -instantaneous evaporative heat flux at skin
-        * ``evap_load_wm2_min`` [W*min/m²] -accumulated evaporative load for chaining
+        * ``sweat_loss_g`` [g] — cumulative sweat mass per person (not area‑normalised)
```
Comment (FedericoTartarini) [inline CHANGELOG.rst:189]: Why did you change this that they are not related to your pull request?

**Review by FedericoTartarini — CHANGES_REQUESTED** [review]
@ChuanLi32781 Great work, thank you very much for fixing the code, I think we are almost there. I've left a few comments that I would like you to review. The main issue at the moment is that the pull request should not be pointing toward master but toward development. Please fix that as soon as possible. and also please review and address all my comments and the comments provided by CodeRabbit. Once you are done, I can review the code again and then help you merge it.

**Comment by FedericoTartarini** [comment]
@ChuanLi32781 Please also fix the merge error with the changelog file.

File: docs/documentation/models.rst (hunk 9 above)
Comment (ChuanLi32781) [reply docs/documentation/models.rst:140]: Removed the unrelated changes from this file and kept only the IREQ-related updates.

Those unrelated changes were originally made while addressing automated review comments.

File: tests/test_ireq.py (hunk 14 above)
Comment (ChuanLi32781) [reply tests/test_ireq.py:4]: I added a new parameterized test based on the ISO 11079 Table F.1 calculation examples.

The test covers all 8 rows from the table and validates IREQneutral, ICLneutral, and DLEneutral against the reported reference values.

File: pythermalcomfort/models/ireq.py [hunk 16]
```diff
@@ -0,0 +1,424 @@
...
+    when available clothing insulation is insufficient, in accordance with
+    ISO 11079 [11079ISO2007]_.
+
+    Parameters
+    ----------
+    tdb : float or list of floats
+        Dry bulb air temperature, [deg C].
+    tr : float or list of floats
+        Mean radiant temperature, [deg C].
+    v : float or list of floats
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py]: if this the relative air speed it should be called vr as in the pmv_ppd_iso

File: pythermalcomfort/models/ireq.py [hunk 17]
```diff
@@ -0,0 +1,424 @@
...
+        Relative air velocity, [m/s].
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py]: it is air speed not velocity

File: pythermalcomfort/models/ireq.py [hunk 18]
```diff
@@ -0,0 +1,424 @@
...
+        ireq_min=results["ireq_min"],
+        ireq_neutral=results["ireq_neutral"],
+        icl_min=results["icl_min"],
+        icl_neutral=results["icl_neutral"],
+        dle_min=results["dle_min"],
+        dle_neutral=results["dle_neutral"],
+    )
+
+
+def _clothing_constant_part(p, v, walk_sp):
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py]: missing docstring

File: pythermalcomfort/models/ireq.py [hunk 19]
```diff
@@ -0,0 +1,424 @@
...
+        & (met <= 400.0)
+        & (tdb <= 10.0)
+        & (v >= 0.4)
+        & (v <= 18.0)
+        & (walk_sp >= minimum_walking_speed)
+        & (walk_sp <= 1.2)
+    )
+
+
+def _format_dle(dle, round_output):
```
Comment (FedericoTartarini) [inline pythermalcomfort/models/ireq.py:408]: missing docstring

**Review by FedericoTartarini — CHANGES_REQUESTED** [review]
Could you please address the open comments?

**Review by FedericoTartarini — CHANGES_REQUESTED** [review]
## Code Review — `ireq` model (ISO 11079)

The implementation is a solid start — the algorithm is numerically careful, the test coverage is reasonable, and the function/class naming (`ireq` / `IREQ`) follows codebase conventions. However there are several issues to address before merging, ranging from a pattern-breaking input class to docstring inconsistencies and a potential numerical bug.

---

### 1. `IREQInputs` breaks the established `BaseInputs` pattern [High]

Every other input class (`PHSInputs`, `HumidexInputs`, `AnkleDraftInputs`, etc.) does exactly one thing: passes arguments up to `super().__init__()` and lets `BaseInputs.__post_init__()` handle type validation. `IREQInputs.__post_init__()` instead performs full numpy broadcasting, array casting, and physical-constraint validation — duplicating work that `BaseInputs` already does and moving logic that belongs in the model function.

```python
# All other input classes follow this pattern:
class PHSInputs(BaseInputs):
    def __init__(self, tdb, tr, v, rh, met, clo, round_output, wme, posture):
        super().__init__(tdb=tdb, tr=tr, ...)
```

**Fix:** `p` and `walk_sp` should be added as fields to `BaseInputs` (as is done for every other field like `v_ankle`, `t_sk`, etc.), and the physical constraint validation (`met > 0`, `v >= 0`, etc.) should move into the model function itself, following the pattern of `humidex.py` or `ankle_draft.py`.

---

### 2. Docstring uses `[deg C]` — should be `[°C]` [Medium]

Every other model uses the Unicode degree symbol. `ireq.py` uses `[deg C]` throughout, which is inconsistent with `phs.py`, `ankle_draft.py`, `humidex.py`, and all other models.

```python
# ireq.py (incorrect):
tdb : float or list of floats
    Dry bulb air temperature, [deg C].

# Correct (matches the rest of the codebase):
tdb : float or list of floats
    Dry bulb air temperature, [°C].
```

---

### 3. Return class type annotations are inconsistent [Medium]

All other return dataclasses use `float | list[float]`. `IREQ` uses `npt.ArrayLike` for most fields, which is inconsistent:

```python
# classes_return.py — IREQ (inconsistent):
ireq_min: npt.ArrayLike
ireq_neutral: npt.ArrayLike

# Every other return class in the same file:
a_pmv: float | list[float]
tmp_cmf: float | list[float]
```

The `dle_*` fields typed as `str | float | np.ndarray` are also unusual. Use `str | float | list` to match user-facing semantics.

---

### 4. `air_insulation` can go negative with `limit_inputs=False` [Medium]

```python
air_insulation = 0.092 * np.exp(-0.15 * v - 0.22 * walk_sp) - 0.0045
```

At v ≈ 22 m/s (above the ISO limit of 18 m/s, but reachable when `limit_inputs=False`), this expression goes negative. Downstream this causes a division by zero in `1.0 / air_insulation` and nonsensical results in `total_evaporative_resistance`. Add a guard or at minimum clip it:

```python
air_insulation = np.maximum(
    0.092 * np.exp(-0.15 * v - 0.22 * walk_sp) - 0.0045,
    1e-6,
)
```

---

### 5. Parameter `p` is cryptically named [Medium]

A single-letter parameter `p` for "Air permeability of clothing" is ambiguous — `p` typically denotes pressure or probability in physics and in this codebase (`p_atm` is already a field in `BaseInputs`). Consider renaming to `p_air`, `air_perm`, or at a minimum adding a prominent note in the docstring that makes the meaning unambiguous.

---

### 6. `clo` description is misleading in IREQ context [Medium]

In every other pythermalcomfort model, `clo` means the clothing the person is actually wearing. In IREQ, `clo` is the clothing *available to* the person — used only in the DLE calculation (second iterative loop) to determine how long they can safely remain in the environment. The docstring just says "Clothing insulation, [clo]" with no explanation of this critical distinction.

---

### 7. Private helpers missing docstrings — CI docstring coverage check fails [Medium]

CodeRabbit's pre-merge check shows docstring coverage at **70%**, below the required **80%** threshold. None of the four private helpers have docstrings:

```python
def _clothing_constant_part(p, v, walk_sp): ...
def _valid_iso_11079_inputs(met, tdb, v, walk_sp): ...
def _format_dle(dle, round_output): ...
def _scalar(value): ...
```

---

### 8. `_clothing_constant_part` called inside the DLE loop unnecessarily [Low]

`_clothing_constant_part(p, v, walk_sp)` is called on every iteration of the second loop even though `p`, `v`, and `walk_sp` are all fixed per call. It should be hoisted above both loops:

```python
# Current (inside loop, up to 150 calls per element):
for _ in range(150):
    ...
    constant_part = _clothing_constant_part(p=p, v=v, walk_sp=walk_sp)

# Fix (compute once):
constant_part = _clothing_constant_part(p=p, v=v, walk_sp=walk_sp)
for _ in range(150):
    ...
```

---

### 9. Magic numbers with no ISO equation references [Low]

The function is dense with unexplained numeric literals (`ar_adu = 0.77`, `29.0`, `0.1333`, `4030.183`, `18.6686`, `235.0`, `0.0354`, `0.0285`, `1.197`, `-40.0`, `0.092`, `5.67e-08`, `0.95`, `1.73e-02`, `1.4e-03`, etc.). Since these come from specific equations in ISO 11079, add inline comments referencing the equation or table they originate from, e.g., `# ISO 11079:2007 Eq. A.5` or `# Table B.1`. This makes validation against the standard tractable for future maintainers.

---

### 10. Test gap: `tdb > 10` out-of-range case not covered [Low]

`test_ireq_limit_inputs_returns_nan_outside_iso_range` only tests `v=0.2` as the out-of-range trigger. The most common real-world out-of-range scenario for a cold-stress model would be `tdb > 10 °C` (the ISO applicability limit). Please add a test case for that boundary.

**Comment by FedericoTartarini** [comment]
@ChuanLi32781 I left a few comments and Claude has also provided some comments. Can you please address them one by one and can you please check the open pending comments? Please reply to all of them and you can mark them as completed if you are sure about the solution that you have implemented, otherwise leave them open and tag me so I can then close them.

**Review by FedericoTartarini — DISMISSED** [review]
